src/boxy.nim

```diff
@@ -361,121 +343,187 @@ proc grow(boxy: Boxy) =
+  let newAllocator = newSkylineAllocator(newAtlasSize, oldAllocator.margin)
 
-  boxy.addWhiteTile()
+  # Create a framebuffer for the new atlas
+  var growFramebufferId: GLuint
+  glGenFramebuffers(1, growFramebufferId.addr)
+  glBindFramebuffer(GL_FRAMEBUFFER, growFramebufferId)
+  glFramebufferTexture2D(
+    GL_FRAMEBUFFER,
+    GL_COLOR_ATTACHMENT0,
+    GL_TEXTURE_2D,
+    newAtlasTexture.textureId,
+    0
+  )
+
+  # Clear the new atlas
+  glViewport(0, 0, newAtlasSize.GLint, newAtlasSize.GLint)
+  glClearColor(0, 0, 0, 0)
+  glClear(GL_COLOR_BUFFER_BIT)
 
-  for y in 0 ..< oldTileRun:
-    for x in 0 ..< oldTileRun:
+  # Set up for drawing to the new atlas
+  let savedProj = boxy.proj
+  let savedMat = boxy.mat
+  let savedMats = boxy.mats
+  boxy.proj = ortho(0.float32, newAtlasSize.float32, 0, newAtlasSize.float32, -1000, 1000)
+  boxy.mat = mat3()
+  boxy.mats = @[]
+
+  # Create new entries table and re-allocate all images
+  var newEntries: Table[string, ImageInfo]
+  for key, oldInfo in oldEntries:
+    if oldInfo.isOneColor:
+      # Solid color images don't need atlas space
+      newEntries[key] = oldInfo
+    else:
+      # Allocate space in the new atlas
+      let allocation = newAllocator.allocate(oldInfo.size.x, oldInfo.size.y)
+      if not allocation.success:
+        raise newException(BoxyError, "Failed to re-allocate image during grow: " & key)
+
+      var newInfo = oldInfo
+      newInfo.atlasPos = ivec2(allocation.x.int32, allocation.y.int32)
+      newEntries[key] = newInfo
+
+      # Draw the image from old position to new position using GPU
       let
-        imageTile = oldAtlas.superImage(
-          x * (boxy.tileSize + tileMargin),
-          y * (boxy.tileSize + tileMargin),
-          boxy.tileSize + tileMargin,
-          boxy.tileSize + tileMargin
-        )
-        index = x + y * oldTileRun
-      updateSubImage(
-        boxy.atlasTexture,
-        (index mod boxy.tileRun) * (boxy.tileSize + tileMargin),
-        (index div boxy.tileRun) * (boxy.tileSize + tileMargin),
-        imageTile
+        srcX = oldInfo.atlasPos.x.float32
+        srcY = oldInfo.atlasPos.y.float32
+        srcW = oldInfo.size.x.float32
+        srcH = oldInfo.size.y.float32
+        dstX = allocation.x.float32
+        dstY = allocation.y.float32
+
+      # Use drawUvRect to copy the image data
+      boxy.drawUvRect(
+        at = vec2(dstX, dstY),
+        to = vec2(dstX + srcW, dstY + srcH),
+        uvAt = vec2(srcX, srcY),
+        uvTo = vec2(srcX + srcW, srcY + srcH),
+        tint = color(1, 1, 1, 1)
       )
 
-proc takeFreeTile(boxy: Boxy): int =
-  let (found, index) = boxy.takenTiles.firstFalse
-  if found:
-    boxy.takenTiles.unsafeSetTrue(index)
-    return index
+  # Flush the draw calls to copy all images
+  boxy.flush()
 
-  boxy.grow()
-  boxy.takeFreeTile()
+  # Restore framebuffer binding
+  if boxy.layerNum >= 0:
+    glBindFramebuffer(GL_FRAMEBUFFER, boxy.layerFramebufferId)
```

Comment:
Can you do the cond only for the 2nd param? saves repeating the whole fn call

Reply:
I am not sure what you mean?

Reply:
glBindFramebuffer(GL_FRAMEBUFFER, if ...: id0 else: id1)